### ISSUE TITLE
CS8602 (dereference of a possible null reference) when using method group, but not when using lambda

### ISSUE THREAD

**Issue by thomaslevesque** [dotnet/roslyn#57753]
**Version Used**: 

C# 10 / .NET SDK 6.0.100

**Steps to Reproduce**:

In a nullable-enabled context, write this code:
```csharp
void TestWithMethodGroup()
{
    var files = Directory.EnumerateFiles(path)
                         .Select(Path.GetFileNameWithoutExtension);
    foreach (var file in files)
    {
        Console.WriteLine(file.Length);
    }
}

void TestWithLambda()
{
    var files = Directory.EnumerateFiles(path)
                         .Select(s => Path.GetFileNameWithoutExtension(s));
    foreach (var file in files)
    {
        Console.WriteLine(file.Length);
    }
}
```

**Expected Behavior**:
CS8602 should not be reported in either of these methods.
`Directory.EnumerateFiles` returns non-nullable strings
`Path.GetFileNameWithoutExtension` returns `string?`, but has a `[return: NotNullIfNotNull("path")]` attribute. Since the input is non-null, the output is known to be non-null as well.

**Actual Behavior**:
In the `TestWithMethodGroup` method, CS8602 is reported on `file.Length`.
In the `TestWithLambda` method, no warning is reported.
Both methods are semantically (if not technically) equivalent, so there's no reason for this difference.

**Comment by RikkiGibson** [comment]
Looks like a duplicate of #44174.

**Comment by RikkiGibson** [comment]
Oops, I closed prematurely. I wanted to give you the opportunity to follow up first.

**Comment by thomaslevesque** [comment]
@RikkiGibson you're right, it's a duplicate. I'll follow the other thread, then. Thanks!
